sync2/p2p.go

```diff
@@ -0,0 +1,211 @@
+func NewP2PHashSync(
+	logger *zap.Logger,
+	d *Dispatcher,
+	name string,
+	os multipeer.OrderedSet,
+	keyLen, maxDepth int,
+	peers *peers.Peers,
+	handler multipeer.SyncKeyHandler,
+	cfg Config,
+	requester rangesync.Requester,
+) *P2PHashSync {
+	s := &P2PHashSync{
+		logger: logger,
+		os:     os,
+		cfg:    cfg,
+	}
+	rangeSyncOpts := []rangesync.RangeSetReconcilerOption{
+		rangesync.WithMaxSendRange(cfg.MaxSendRange),
+		rangesync.WithSampleSize(cfg.SampleSize),
+		rangesync.WithMaxDiff(cfg.MaxReconcDiff),
+		rangesync.WithLogger(logger),
+	}
+	if cfg.RecentTimeSpan > 0 {
+		rangeSyncOpts = append(rangeSyncOpts, rangesync.WithRecentTimeSpan(cfg.RecentTimeSpan))
+	}
+	// var ps multipeer.PairwiseSyncer
+	ps := rangesync.NewPairwiseSetSyncer(requester, name, rangeSyncOpts, []rangesync.ConduitOption{
+		rangesync.WithTrafficLimit(cfg.TrafficLimit),
+		rangesync.WithMessageLimit(cfg.MessageLimit),
+	})
+	s.syncBase = multipeer.NewSetSyncBase(ps, s.os, handler)
+	s.reconciler = multipeer.NewMultiPeerReconciler(
+		s.syncBase, peers, keyLen, maxDepth,
+		multipeer.WithLogger(logger),
+		multipeer.WithSyncPeerCount(cfg.SyncPeerCount),
+		multipeer.WithMinSplitSyncPeers(cfg.MinSplitSyncPeers),
+		multipeer.WithMinSplitSyncCount(cfg.MinSplitSyncCount),
+		multipeer.WithMaxFullDiff(cfg.MaxFullDiff),
+		multipeer.WithSyncInterval(cfg.SyncInterval),
+		multipeer.WithMinCompleteFraction(cfg.MinCompleteFraction),
+		multipeer.WithSplitSyncGracePeriod(time.Minute),
+		multipeer.WithNoPeersRecheckInterval(cfg.NoPeersRecheckInterval))
+	d.Register(name, s.serve)
+	return s
+}
+
+func (s *P2PHashSync) serve(ctx context.Context, stream io.ReadWriter) error {
+	peer, found := server.ContextPeerID(ctx)
+	if !found {
+		panic("BUG: no peer ID found in the handler")
+	}
+	// We derive a dedicated Syncer for the peer being served to pass all the received
+	// items through the handler before adding them to the main ItemStore
+	return s.syncBase.Derive(peer).Serve(ctx, stream)
+}
+
+// Set returns the OrderedSet that is being synchronized.
+func (s *P2PHashSync) Set() rangesync.OrderedSet {
+	return s.os
+}
+
+// Load loads the OrderedSet from the underlying storage.
+func (s *P2PHashSync) Load() error {
+	s.logger.Info("loading the set")
+	start := time.Now()
+	// We pre-load the set to avoid waiting for it to load during a
+	// sync request
+	if err := s.os.EnsureLoaded(); err != nil {
+		return fmt.Errorf("load set: %w", err)
+	}
+	info, err := s.os.GetRangeInfo(nil, nil)
+	if err != nil {
+		return fmt.Errorf("get range info: %w", err)
+	}
+	s.logger.Info("done loading the set",
+		zap.Duration("elapsed", time.Since(start)),
+		zap.Int("count", info.Count),
+		zap.Stringer("fingerprint", info.Fingerprint))
+	return nil
+}
+
+// Start starts the multi-peer reconciler.
+func (s *P2PHashSync) Start() {
+	if !s.cfg.EnableActiveSync {
+		s.logger.Info("active sync is disabled")
+		return
+	}
+	s.running.Store(true)
+	s.start.Do(func() {
```

Comment:
not sure why `Once` is needed? can it be that it will be called from multiple places?

Reply:
The idea was for the `Start` method to be idempotent

Reply:
I don't quite understand the reasoning - `Start` can be called any amount of times - but only the first time it actually does something? `Start` - `Stop` - `Start` causes `running` to be `true` but the component is actually in the stopped state?

Reply:
The idea is that there's no harm in invoking `Start` multiple times on a `P2PHashSync`, but after you `Stop()` it, you throw it away (it is non-restartable)

Reply:
For me the interesting question is why is it possible in the first place that `Start` would be called multiple times

Reply:
This will be more evident from forthcoming ATX sync integration PR.
Basically "ensure the syncer is active" is what's needed there. W/o this `Once`, we'd need to keep track of whether the particular syncer is already started